### PR TITLE
Max attempt retry connection to cold db

### PR DESCRIPTION
**Pull request by epifanio**
# Overview
This PR attempts to mitigate a connection issues which can arise when the service attempts to connecting to a 'cold-db' (e.g.: service in idle state in a docker swarm setting). The retrty loop uses a new configure option `max_db_conn_retry_attempt` to retry the connection until  such value is reached - default value for `max_db_conn_retry_attempt` is set to `5`.

# Related Issue / Discussion

# Additional Information

# Contributions and Licensing

(as per https://github.com/geopython/pycsw/blob/master/CONTRIBUTING.rst#contributions-and-licensing)

- [x] I'd like to contribute [feature X|bugfix Y|docs|something else] to pycsw. I confirm that my contributions to pycsw will be compatible with the pycsw license guidelines at the time of contribution.
- [x] I have already previously agreed to the pycsw Contributions and Licensing Guidelines

### CODE REVIEW

**Comment by epifanio** [comment]
cc @tomkralidis , @kalxas , @ricardogsilva  I tried to rewrite this PR from a clean branch as the last one got messed-up by a wrong merge with my fork made by mistake.

**Review by tomkralidis — CHANGES_REQUESTED** [review]

File: default-sample.cfg [hunk 1]
```diff
@@ -45,6 +45,7 @@ gzip_compresslevel=9
 #spatial_ranking=true
 profiles=apiso
 #workers=2
+#max_db_conn_retry_attempt=5
```
[the remark this answers is not in the document]
Comment (epifanio) [reply default-sample.cfg]: changed in https://github.com/epifanio/pycsw/commit/b3d898c18ffc97170b40723f87569802f34ee7ba

File: docs/configuration.rst [hunk 2]
```diff
@@ -25,6 +25,7 @@ pycsw's runtime configuration is defined by ``default.cfg``.  pycsw ships with a
 - **smtp_host**: SMTP host for processing ``csw:ResponseHandler`` parameter via outgoing email requests (default is ``localhost``)
 - **spatial_ranking**: parameter that enables (``true`` or ``false``) ranking of spatial query results as per `K.J. Lanfear 2006 - A Spatial Overlay Ranking Method for a Geospatial Search of Text Objects  <http://pubs.usgs.gov/of/2006/1279/2006-1279.pdf>`_.
 - **workers**: set the number of workers used by the wsgi server when lunching pycsw using the provided docker/entrypoint.py. If not set, it will use 2 workers as Default.
+- **max_db_conn_retry_attempt**: max number of retry attempts when connecting to records-repository database
```
[the remark this answers is not in the document]
Comment (epifanio) [reply docs/configuration.rst]: applied in https://github.com/epifanio/pycsw/commit/b3d898c18ffc97170b40723f87569802f34ee7ba

**Review by tomkralidis — APPROVED** [review]
LGTM cc @kalxas

**Comment by epifanio** [comment]
> LGTM cc @kalxas

@kalxas @tomkralidis  something I need to fix?

**Comment by tomkralidis** [comment]
@epifanio it looks like CI is timing out.  I'm wondering whether the `while` loops in `pycsw/server.py` never evaluate to `True`?

I looked back at your previous commits in this PR and have below an snippet which may help (needs testing):

```diff
diff --git a/pycsw/server.py b/pycsw/server.py
index 3f26cc0c..1ce23edd 100644
--- a/pycsw/server.py
+++ b/pycsw/server.py
@@ -205,6 +205,12 @@ class Csw(object):
                     'NoApplicableCode', 'service',
                     'Could not load repository.mappings')
 
+        # load user-defined max attempt to retry db connection
+        try:
+            self.max_retries = int(self.config.get("repository", "max_retries"))
+        except configparser.NoOptionError:
+            self.max_retries = 5
+
         # load outputschemas
         LOGGER.info('Loading outputschemas')
 
@@ -399,8 +405,18 @@ class Csw(object):
             rs_cls = getattr(rs_mod, rs_clsname)
 
             try:
-                self.repository = rs_cls(self.context, repo_filter)
-                LOGGER.debug('Custom repository %s loaded (%s)', rs, self.repository.dbtype)
+                connection_done = False
+                max_attempts = 0
+
+                while not connection_done and max_attempts <= 5:
+                    try:
+                        self.repository = rs_cls(self.context, repo_filter)
+                        LOGGER.debug('Custom repository %s loaded (%s)', rs, self.repository.dbtype)
+                        connection_done = True
+                    except:
+                        LOGGER.debug(f'Repository not loaded retry connection {max_attempts}')
+                        max_attempts += 1
+
             except Exception as err:
                 msg = 'Could not load custom repository %s: %s' % (rs, err)
                 LOGGER.exception(msg)
@@ -414,15 +430,26 @@ class Csw(object):
             from pycsw.core import repository
             try:
                 LOGGER.info('Loading default repository')
-                self.repository = repository.Repository(
-                    self.config.get('repository', 'database'),
-                    self.context,
-                    self.environ.get('local.app_root', None),
-                    self.config.get('repository', 'table'),
-                    repo_filter
-                )
-                LOGGER.debug(
-                    'Repository loaded (local): %s.' % self.repository.dbtype)
+
+                connection_done = False
+                max_attempts = 0
+
+                while not connection_done and max_attempts <= self.max_retries:
+                    try:
+                        self.repository = repository.Repository(
+                            self.config.get('repository', 'database'),
+                            self.context,
+                            self.environ.get('local.app_root', None),
+                            self.config.get('repository', 'table'),
+                            repo_filter
+                        )
+                        LOGGER.debug(
+                            'Repository loaded (local): %s.' % self.repository.dbtype)
+                        connection_done = True
+                    except:
+                        LOGGER.debug(f'Repository not loaded retry connection {max_attempts}')
+                        max_attempts += 1
+
             except Exception as err:
                 msg = 'Could not load repository (local): %s' % err
                 LOGGER.exception(msg)
```

**Pull request merged**